libexec/cli/build.exec

```diff
@@ -250,9 +250,29 @@ case $SINGULARITY_BUILDDEF in
         message 1 "Importing: base Singularity environment\n"
         zcat "$SINGULARITY_libexecdir/singularity/bootstrap-scripts/environment.tar" | tar xBf - -C "${SINGULARITY_ROOTFS}" || exit $?
 
+        # Docker layers are extracted with tar, but we need additional handling
+        # for aufs whiteout files that may be present.
+        # - .wh..whi..opq inside a directory indicates that directory is opaque.
+        #    Any content in this directory from previous layers must be deleted
+        #    before extracting the current layer.
+        # - .wh.<file/dirname> indicates <file/dirname> from a previous layer must
+        #    be deleted as we extract the current layer.
+        # See:
+        #   https://github.com/tonistiigi/docker/blob/2fb5d0c32376951ef41a6f64bb7dbd8f6fd14fba/pkg/archive/whiteouts.go#L3-L23
         for i in `cat "$SINGULARITY_CONTENTS"`; do
             message 1 "Importing: $i\n"
-            zcat "$i" | (cd "$SINGULARITY_ROOTFS"; tar --exclude=dev/* --exclude=*/.wh.* -xf - ) || exit $?
+            # Remove any existing directories containing a whiteout opaque
+            # indicator file .wh..wh..opq in the layer we are about to extract
+            zcat "$i" | (cd "$SINGULARITY_ROOTFS"; tar -tf - | grep '\.wh\.\.wh\.\.opq' | sed 's/\.wh\.\.wh\.\.opq//' | xargs -r -d '\n' echo Opaque: rm -rf ) || exit $?
+            zcat "$i" | (cd "$SINGULARITY_ROOTFS"; tar -tf - | grep '\.wh\.\.wh\.\.opq' | sed 's/\.wh\.\.wh\.\.opq//' | xargs -r -d '\n' rm -rf ) || exit $?
+            # Now extract our current layer, exclude whiteout markes handled
+            # above.
+            zcat "$i" | (cd "$SINGULARITY_ROOTFS"; tar --exclude=dev/* --exclude=*/.wh..wh..opq -xf - ) || exit $?
+            # Get rid of any files/dirs marked for white out
+            find "$SINGULARITY_ROOTFS" -type f -name '.wh.*' -print | sed 's/\.wh\.//' | xargs -r -d '\n' echo Whiteout: rm -rf
+            find "$SINGULARITY_ROOTFS" -type f -name '.wh.*' -print | sed 's/\.wh\.//' | xargs -r -d '\n' rm -rf
+            # Get rid of all of the whiteout indicator files themselves
+            find "$SINGULARITY_ROOTFS" -type f -name '.wh.*' -delete
         done
```

Comment:
holy cow, this is some serious bash-fu. I'll need to look this over this carefully tomorrow! I still need to read more about this, I don't understand how we can know about whiteouts in a layer above/below without looking at one folder across all the layers at once. For example, if you find `.wh.wh.opq.folder` in layer 1, but then layer 2 doesn't have that folder whited out, how would we remember the context from the first?

Reply:
See my comment above - this *is* kinda crazy shell fu. It's also not particularly safe.

AFAIK whiteouts don't need to be remembered across all layers. You only care about your current layer vs what you have extracted up to that point.

Opaque folders occur when you delete a folder, then recreate it. `rm -r bob; mkdir bob`. A whiteout file would mean bob is not present (remvoved) in the current layer. We did remove it, but then recreated it. The opaque marker file shows yes, bob does exist but we don't want anything that used to be there to show through any more.